ctru-rs/src/lib.rs

```diff
@@ -38,12 +38,31 @@ pub fn init() {
         // Only for panics in the main thread
         if main_thread == thread::current().id() && console::Console::exists() {
             println!("\nPress SELECT to exit the software");
-            let hid = services::hid::Hid::init().unwrap();
 
-            loop {
-                hid.scan_input();
-                if hid.keys_down().contains(services::hid::KeyPad::KEY_SELECT) {
-                    break;
+            // The use of unsafe functions here is basically obligatory.
+            // To have memory safety when using the `Hid` struct, we must not make more
+            // than one available at the same time, so no drop/service ownership issues arise.
+            // The problem here is that the `panic_hook` is run _before_ the app cleanup,
+            // so an `Hid` stuct may still be alive and thus make the `panic_hook` panic.
+            // If that were to happen, the system would have to reboot to properly close the app.
+            // 
+            // Using `hidInit` is safe when another instance is open, and we can do safe operations afterwards.
+            // The only (probably) unsafe part of this system is the `hidExit`, since in a multithreaded
+            // environment some other threads may still be doing operations on the service
+            // before the cleanup, though the time window would be almost nonexistent, and it would only
+            // really be a problem in preemptive threads.
+            //
+            // TL;DR : This code is bad.
```

Comment:
Maybe we need to do something a little more fancy than an atomic boolean.

One idea is having a static `Mutex<usize>` which starts at zero and counts the number of open service handles. When it reaches zero again in drop, we can call the exit function. It needs to be a Mutex and not just an AtomicUsize because we need to hold the lock while we call the `hidExit` function, for example. Otherwise, we might decrement to zero, another thread calls `Hid::init` and increments to one (calling `hidInit`), and the drop function finally calls `hidExit`.

This would let us allow multiple handles to be created without issue, and only exiting the service once all are gone. Basically re-implementing reference counting, but it's a little hard to reuse `Arc` for this (feel free to try if you want, but it will probably be more verbose and complex).

We could code this pattern up into a `ServiceCleaner` or something (name tbd), like so (untested):
```rust
struct ServiceCleaner {
    counter: &'static Mutex<usize>
    exit: extern "C" fn()
}

impl ServiceCleaner {
    fn init(counter: &'static Mutex<usize>, exit: extern "C" fn()) -> Self {
        let mut value = counter.lock().unwrap() // todo: handle poisoning
        *value += 1;
        drop(value);
        Self { counter, exit }
    }
}

impl Drop for ServiceCleaner {
    fn drop(&mut self) {
        let mut value = self.counter.lock().unwrap(); // should probably handle poisoning - could just map_err to ignore it.
        *value -= 1;
        if *value == 0 {
            unsafe { self.exit() };
        }
    }
}

```

Reply:
One place to be careful about with the above is `Gfx`. We should probably still only let them have one instance of that struct.

Reply:
`Gfx` absolutely must not be re-initialised, since it takes parameters that change the internal state. Other services, like `Hid` or `SslC` have no problem being shared across threads, though we should be careful with what the services do.

Reply:
Regarding the function naming, `init` is probably fine so long as the underlying ctru_sys funciton (ex. `aptInit`) is OK if it's called multiple times.

I checked all of the "*Init()" functions in libctru, and they all\* actually handle init/exit with reference counting correctly. So we could don't need to do anything special. Even the `gfxInit` function is partially guarded (the `gspInit` call only inits once, but the rest of the gfx settings get reset on every call).

\* romfsInit does try to mount it every time, so it might error out. And we still may still want to limit the Gfx initialization.